Update fixtures / transaction tests

cburgdorf opened the pull request:
### What was wrong?

The fixtures where outdated. (See #208 

### How was it fixed?

I tried to address #208 and updated the fixtures and tried to refactor the transaction tests to match the new structure. However, I feel like I might be missing something and could use some guidance. The main thing I'm wondering about goes as follows.

The old transaction fixtures looked like this.


[TransactionTests/ttFrontier/TransactionWithGasLimitOverflow63_1.json](https://github.com/ethereum/tests/blob/6df9b4ad039ff0043ca75579979f945598898eb3/TransactionTests/ttFrontier/TransactionWithGasLimitOverflow63_1.json)
```json
{
    "TransactionWithGasLimitOverflow63_1" : {
        "_info" : {
            "comment" : "",
            "filledwith" : "cpp-1.3.0+commit.6e0ce939.Linux.g++",
            "lllcversion" : "Version: 0.4.18-develop.2017.10.6+commit.6c09e32c.Linux.g++",
            "source" : "src/TransactionTestsFiller/ttFrontier/TransactionWithGasLimitOverflow63_1Filler.json",
            "sourceHash" : "e481f3dbba3a185bf2a937913e0d90a40960b9e4dcbbfc929870e328c46473f5"
        },
        "blocknumber" : "0",
        "hash" : "f4a1caa6a8a3c08b8fe1528c862e626720298edb7dca2aee3d10ff8a947af12e",
        "rlp" : "0xf865800188800000000000000194095e7baea6a6c7c4c2dfeb977efac326af552d870b801ba048b55bfa915ac795c431978d8a6a992b628d557da5ff759b307d495a36649353a0efffd310ac743f371de3b9f7f9cb56c0b28ad43601b4ab949f53faa07bd2c804",
        "sender" : "41cc818b2cd4606ff8981a38daf7c4f12d40ed28",
        "transaction" : {
            "data" : "",
            "gasLimit" : "0x8000000000000001",
            "gasPrice" : "0x01",
            "nonce" : "0x00",
            "r" : "0x48b55bfa915ac795c431978d8a6a992b628d557da5ff759b307d495a36649353",
            "s" : "0xefffd310ac743f371de3b9f7f9cb56c0b28ad43601b4ab949f53faa07bd2c804",
            "to" : "0x095e7baea6a6c7c4c2dfeb977efac326af552d87",
            "v" : "0x1b",
            "value" : "0x0b"
        }
    }
```
The same fixture now looks like this.

[TransactionTests/ttGasLimit/TransactionWithGasLimitOverflow63_1.json](https://github.com/ethereum/tests/blob/47b09f42c0681548a00da5ab1c98808b368af49a/TransactionTests/ttGasLimit/TransactionWithGasLimitOverflow63_1.json)

```json
{
    "TransactionWithGasLimitOverflow63_1" : {
        "Byzantium" : {
            "hash" : "2a9ef120d86d735b6b7b2a43462d2567a4e1605d18fa5b3c70cb956cbcb54e3a",
            "sender" : "cda94a77b8357ff651db2e3828216440c4b732e8"
        },
        "Constantinople" : {
            "hash" : "2a9ef120d86d735b6b7b2a43462d2567a4e1605d18fa5b3c70cb956cbcb54e3a",
            "sender" : "cda94a77b8357ff651db2e3828216440c4b732e8"
        },
        "EIP150" : {
            "hash" : "2a9ef120d86d735b6b7b2a43462d2567a4e1605d18fa5b3c70cb956cbcb54e3a",
            "sender" : "cda94a77b8357ff651db2e3828216440c4b732e8"
        },
        "EIP158" : {
            "hash" : "2a9ef120d86d735b6b7b2a43462d2567a4e1605d18fa5b3c70cb956cbcb54e3a",
            "sender" : "cda94a77b8357ff651db2e3828216440c4b732e8"
        },
        "Frontier" : {
            "hash" : "2a9ef120d86d735b6b7b2a43462d2567a4e1605d18fa5b3c70cb956cbcb54e3a",
            "sender" : "cda94a77b8357ff651db2e3828216440c4b732e8"
        },
        "Homestead" : {
            "hash" : "2a9ef120d86d735b6b7b2a43462d2567a4e1605d18fa5b3c70cb956cbcb54e3a",
            "sender" : "cda94a77b8357ff651db2e3828216440c4b732e8"
        },
        "_info" : {
            "comment" : "",
            "filledwith" : "cpp-1.3.0+commit.1829957d.Linux.g++",
            "lllcversion" : "Version: 0.4.18-develop.2017.10.11+commit.81f9f86c.Linux.g++",
            "source" : "src/TransactionTestsFiller/ttGasLimit/TransactionWithGasLimitOverflow63_1Filler.json",
            "sourceHash" : "7673d796a4a614c533ac61ecb0f8c49158e06d8b86932813afd3d0be13ae56e4"
        },
        "rlp" : "0xf865800188800000000000000194095e7baea6a6c7c4c2dfeb977efac326af552d870b801ba048b55bfa915ac795c431978d8a6a992b628d557da5ff759b307d495a36649353a01fffd310ac743f371de3b9f7f9cb56c0b28ad43601b4ab949f53faa07bd2c804"
    }
}
```

There's now `hash` and `sender` for each fork but the actual `transaction` data is gone which I'm not entirely sure how to deal with.

I'll leave a couple more inline comments.

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://images.unsplash.com/photo-1452721226468-f95fb66ebf83?ixlib=rb-0.3.5&ixid=eyJhcHBfaWQiOjEyMDd9&s=ecd671f71cc9a6a8a4eb10c7500de2cf&auto=format&fit=crop&w=2000&q=80)

cburgdorf commented on tests/json-fixtures/test_transactions.py:
```diff
@@ -40,43 +57,91 @@
+    expanded one step further to have one fixture for each defined fork within
+    the fixture.
+    """
+    # TODO: This was copied from state tests. Make more DRY.
```
I'll plan to clean this up to have less duplicated code

cburgdorf commented on tests/json-fixtures/test_transactions.py:
```diff
@@ -40,43 +57,91 @@
-    TransactionClass = vm.get_transaction_class()
+def test_transaction_fixtures(fixture, fixture_transaction_class):
+
+    #TODO: Clarify the following:
```
@pipermerriam mind chiming in here to explain how to deal with the missing block number?

cburgdorf commented on tests/json-fixtures/test_transactions.py:
```diff
@@ -40,43 +57,91 @@
         transaction = rlp.decode(fixture['rlp'], sedes=TransactionClass)
-        expected = normalize_signed_transaction(fixture['transaction'])
-
-        assert transaction.nonce == expected['nonce']
```
@pipermerriam as I wrote in the PR description. The new fixture data doesn't have the `transaction` data anymore. Or maybe I didn't understand how to put the pieces together?

cburgdorf replied:
Yep, I was thinking that basically we are just validating the `rlp` with the new tests. That said, there are still a couple of exceptions that we do need to catch and assert correctly (e.g. invalid RLP cases).

I must admit, I'm not entirely sure about the role of the `hash` in the fixtures. Does an existing `hash` just indicate that a transaction is valid whereas the absence of a `hash` in a transaction means it is invalid?

I also think that we do want to keep this part, no?

```
try:
    txn.validate()
except ValidationError:
    return
```

cburgdorf commented on tests/json-fixtures/test_transactions.py:
```diff
@@ -40,43 +45,92 @@
+
+def test_transaction_fixtures(fixture, fixture_transaction_class):
+
+    # TODO: Clarify the following:
```
@pipermerriam previous comment got lost by my latest force push so I'm readding. Mind chiming in here to comment on my comment about the missing block number?

pipermerriam commented:
> the actual transaction data is gone which I'm not entirely sure how to deal with.

I'm pretty sure `rlp` field represents the rlp encode transaction.  you just need to do something like `rlp.decode(fixture[rlp], sedes=vm.get_transaction_class())`

cburgdorf commented on eth_typing/enums.py:
```diff
@@ -0,0 +1,12 @@
+from enum import Enum
+
+
+class ForkName(Enum):
```
uh, interesting. Kinda makes `enum` pointless for string enums then I guess.

cburgdorf commented on tests/json-fixtures/test_transactions.py:
```diff
@@ -40,43 +45,92 @@
+    for fixture_path, fixture_key in all_fixtures:
+        fixture = load_fixture(fixture_path, fixture_key)
+        for fixture_fork, fork_states in fixture.items():
+            if (fixture_fork in ForkName.__members__):
```
That's a good point. The reason I did that is because the fork names are on the same level as `rlp` and `info` but I agree to your point so it would actually be safer to just skip `_info` and `rlp` but let everything else go through so that we won't miss a fork test.

cburgdorf commented:
@pipermerriam Thanks for leaving your comments. I think I addressed your points and updated the PR. I also introduced the changes to the build to falke8/mypy check the new `eth_typing` lib that this PR introduces.

pipermerriam commented on tests/json-fixtures/test_transactions.py:
```diff
@@ -40,68 +45,90 @@
+        try:
+            assert txn.get_sender() == sender
+        except BadSignature:
+            assert not (27 <= txn.v <= 34)
```
Can you explain this assertion.  I don't exactly understand what it's validating or why it's here.

cburgdorf replied:
I can not :) I just kept this one as [we currently have it](https://github.com/ethereum/py-evm/blob/5c273fff1ed8f4815b7232a442061673c00f1c60/tests/json-fixtures/test_transactions.py#L86). I can take it out.

cburgdorf replied:
Git blame says @DavidKnott [would be the one who could answer it](https://github.com/ethereum/py-evm/commit/f868a6e47256eb334fdb1ed620604ee9390e1af0) ;)

pipermerriam replied:
lol, maybe see if removing it causes failures?  This looks like it's an [EIP155](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-155.md) related thing, but I don't think that those should cause bad signature failures but I'm not sure about that.  Maybe check what go-ethereum is doing?

pipermerriam commented on tests/json-fixtures/test_transactions.py:
```diff
@@ -40,68 +45,90 @@
+        sender = to_canonical_address(fixture['sender'])
+
+        try:
+            assert txn.get_sender() == sender
```
No need to change this here but I wanted to be sure you knew about [`eth_utils.is_same_address`](https://github.com/ethereum/eth-utils#is_same_addressa-b---bool) for cases where you need to compare addresses in different forms.

cburgdorf replied:
Thanks for bringing that up!

cburgdorf replied:
I actually updated the code to use it while I was already at it.

pipermerriam commented on tests/json-fixtures/test_transactions.py:
```diff
@@ -40,68 +45,90 @@
+    except (rlp.DeserializationError, rlp.exceptions.DecodingError):
+        assert 'hash' not in fixture, "Transaction was supposed to be valid"
+    except TypeError as err:
+        assert err.args == ("'bytes' object cannot be interpreted as an integer",)
```
Can you add a comment which indicates what this is caused by?

cburgdorf replied:
Yep! Updated

pipermerriam commented on tests/json-fixtures/test_transactions.py:
```diff
@@ -40,68 +41,92 @@
 
 
+# Fixtures have an `_info` key at their root which we need to skip over.
+FIXTURE_FORK_SKIPS = {'_info', 'rlp'}
```
@cburgdorf the `_info` keys makes sense.  Can you fill me in on why the `'rlp'` key is here?

cburgdorf replied:
The forks are on the same level in the fixtures as `_info` and `rlp` so I'm just iterating over all keys and assume everything but `_info` and `rlp` is holding fork specific data.

```
{
    "TransactionWithGasLimitOverflow63_1" : {
        "Byzantium" : {
            "hash" : "2a9ef120d86d735b6b7b2a43462d2567a4e1605d18fa5b3c70cb956cbcb54e3a",
            "sender" : "cda94a77b8357ff651db2e3828216440c4b732e8"
        },
        "Constantinople" : {
            "hash" : "2a9ef120d86d735b6b7b2a43462d2567a4e1605d18fa5b3c70cb956cbcb54e3a",
            "sender" : "cda94a77b8357ff651db2e3828216440c4b732e8"
        },
        "EIP150" : {
            "hash" : "2a9ef120d86d735b6b7b2a43462d2567a4e1605d18fa5b3c70cb956cbcb54e3a",
            "sender" : "cda94a77b8357ff651db2e3828216440c4b732e8"
        },
        "EIP158" : {
            "hash" : "2a9ef120d86d735b6b7b2a43462d2567a4e1605d18fa5b3c70cb956cbcb54e3a",
            "sender" : "cda94a77b8357ff651db2e3828216440c4b732e8"
        },
        "Frontier" : {
            "hash" : "2a9ef120d86d735b6b7b2a43462d2567a4e1605d18fa5b3c70cb956cbcb54e3a",
            "sender" : "cda94a77b8357ff651db2e3828216440c4b732e8"
        },
        "Homestead" : {
            "hash" : "2a9ef120d86d735b6b7b2a43462d2567a4e1605d18fa5b3c70cb956cbcb54e3a",
            "sender" : "cda94a77b8357ff651db2e3828216440c4b732e8"
        },
        "_info" : {
            "comment" : "",
            "filledwith" : "cpp-1.3.0+commit.1829957d.Linux.g++",
            "lllcversion" : "Version: 0.4.18-develop.2017.10.11+commit.81f9f86c.Linux.g++",
            "source" : "src/TransactionTestsFiller/ttGasLimit/TransactionWithGasLimitOverflow63_1Filler.json",
            "sourceHash" : "7673d796a4a614c533ac61ecb0f8c49158e06d8b86932813afd3d0be13ae56e4"
        },
        "rlp" : "0xf865800188800000000000000194095e7baea6a6c7c4c2dfeb977efac326af552d870b801ba048b55bfa915ac795c431978d8a6a992b628d557da5ff759b307d495a36649353a01fffd310ac743f371de3b9f7f9cb56c0b28ad43601b4ab949f53faa07bd2c804"
    }
}
```

cburgdorf replied:
As an alternative we could also do a second check to see if the key is actually an object or a primitive? I just went for a basic blacklist right away ;)

pipermerriam replied:
Lets change this to use the following logic.

- if the key is `_info` or `rlp`, skip
- if the key is in the list of known fork names, yield
- else throw error.

This will guarantee we don't silently skip something at a later date if the format changes.

cburgdorf replied:
We basically do that already because any key that is not in the list of know fork names will throw [here](https://github.com/cburgdorf/py-evm/blob/e8e86e5bbb5ba7aca03d5f3cfbd4c08bf14f3537/tests/json-fixtures/test_transactions.py#L102)
